File: src/main/java/org/jdc/template/event/RxBus.java
package org.jdc.template.event;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

import javax.annotation.Nonnull;
import javax.inject.Inject;
import javax.inject.Singleton;

import rx.Observable;
import rx.Scheduler;
import rx.Subscription;
import rx.android.schedulers.AndroidSchedulers;
import rx.functions.Action1;
import rx.schedulers.Schedulers;
import rx.subjects.PublishSubject;
import rx.subjects.SerializedSubject;
import rx.subjects.Subject;

@Singleton
public class RxBus {
    private final Subject<Object, Object> bus = new SerializedSubject<>(PublishSubject.create());

    private Map<Class, Object> stickyEventList = new ConcurrentHashMap<>();

    @Inject
    public RxBus() {
    }

    public void post(Object o) {
        bus.onNext(o);
    }

    public void postSticky(Object o) {
        stickyEventList.put(o.getClass(), o);
        post(o);
    }

    public void removeSticky(Class eventClass) {
        stickyEventList.remove(eventClass);
    }

    public Observable<Object> toObserverable() {
        return bus;
    }

    public final Subscription subscribeMainThread(@Nonnull Action1<? super Object> onNext) {
        Subscription subscription = subscribe(AndroidSchedulers.mainThread(), onNext);
        postStickyEvents();
        return subscription;
    }

    public final Subscription subscribeBackgroundThread(@Nonnull Action1<? super Object> onNext) {
        Subscription subscription = subscribe(Schedulers.newThread(), onNext);
        postStickyEvents();
        return subscription;
    }

    public Subscription subscribe(@Nonnull Scheduler scheduler, @Nonnull Action1<? super Object> onNext) {
        return bus.observeOn(scheduler).subscribe(onNext);
    }

    private void postStickyEvents() {
        for (Object event : stickyEventList.values()) {
            post(event);
        }
    }
}

File: src/main/kotlin/org/jdc/template/event/RxBus.kt
package org.jdc.template.event

import rx.Observable
import rx.Scheduler
import rx.Subscription
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import rx.subjects.PublishSubject
import rx.subjects.SerializedSubject
import java.util.concurrent.ConcurrentHashMap
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class RxBus
@Inject
constructor() {
    private val bus = SerializedSubject(PublishSubject.create<Any>())

    private val stickyEventList = ConcurrentHashMap<Class<Any>, Any>()

    fun post(o: Any) {
        bus.onNext(o)
    }

    fun postSticky(o: Any) {
        stickyEventList.put(o.javaClass, o)
        post(o)
    }

    fun removeSticky(eventClass: Class<Any>) {
        stickyEventList.remove(eventClass)
    }

    fun toObserverable(): Observable<Any> {
        return bus
    }

    fun subscribeMainThread(onNext: (Any) -> Unit): Subscription {
        val subscription = subscribe(AndroidSchedulers.mainThread(), onNext)
        postStickyEvents()
        return subscription
    }

    fun subscribeBackgroundThread(onNext: (Any) -> Unit): Subscription {
        val subscription = subscribe(Schedulers.newThread(), onNext)
        postStickyEvents()
        return subscription
    }

    fun subscribe(scheduler: Scheduler, onNext: (Any) -> Unit): Subscription {
        return bus.observeOn(scheduler).subscribe(onNext)
    }

    private fun postStickyEvents() {
        for (event in stickyEventList.values) {
            post(event)
        }
    }
}
